lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,1115 @@
+use crate::util::ser::{HighZeroBytesDroppedBigSize, SeekReadable, WithoutLength, Writeable, Writer};
+use crate::util::string::PrintableString;
+
+use crate::prelude::*;
+
+const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "signature");
+
+/// Builds an [`InvoiceRequest`] from an [`Offer`] for the "offer to be paid" flow.
+///
+/// See [module-level documentation] for usage.
+///
+/// [module-level documentation]: self
+pub struct InvoiceRequestBuilder<'a> {
+	offer: &'a Offer,
+	invoice_request: InvoiceRequestContents,
+}
+
+impl<'a> InvoiceRequestBuilder<'a> {
+	pub(super) fn new(offer: &'a Offer, metadata: Vec<u8>, payer_id: PublicKey) -> Self {
+		Self {
+			offer,
+			invoice_request: InvoiceRequestContents {
+				payer: PayerContents(metadata), offer: offer.contents.clone(), chain: None,
+				amount_msats: None, features: InvoiceRequestFeatures::empty(), quantity: None,
+				payer_id, payer_note: None,
+			},
+		}
+	}
+
+	/// Sets the [`InvoiceRequest::chain`] of the given [`Network`] for paying an invoice. If not
+	/// called, [`Network::Bitcoin`] is assumed. Must be supported by the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn chain(mut self, network: Network) -> Self {
+		self.invoice_request.chain = Some(ChainHash::using_genesis_block(network));
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::amount_msats`] for paying an invoice. Must be at least the base
+	/// invoice amount (i.e., [`Offer::amount`] times [`quantity`]).
+	///
+	/// Successive calls to this method will override the previous setting.
+	///
+	/// [`quantity`]: Self::quantity
+	pub fn amount_msats(mut self, amount_msats: u64) -> Self {
+		self.invoice_request.amount_msats = Some(amount_msats);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::features`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	#[cfg(test)]
+	pub fn features(mut self, features: InvoiceRequestFeatures) -> Self {
+		self.invoice_request.features = features;
+		self
+	}
+
+	/// Sets [`InvoiceRequest::quantity`] of items. If not set, `1` is assumed. Must conform to
+	/// [`Offer::is_valid_quantity`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn quantity(mut self, quantity: u64) -> Result<Self, SemanticError> {
+		self.invoice_request.offer.check_quantity(Some(quantity))?;
+		self.invoice_request.quantity = Some(quantity);
+		Ok(self)
+	}
+
+	#[cfg(test)]
+	pub fn quantity_unchecked(mut self, quantity: u64) -> Self {
+		self.invoice_request.quantity = Some(quantity);
+		self
+	}
+
+	/// Sets the [`InvoiceRequest::payer_note`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn payer_note(mut self, payer_note: String) -> Self {
+		self.invoice_request.payer_note = Some(payer_note);
+		self
+	}
+
+	/// Builds an [`InvoiceRequest`] after checking for valid semantics.
```

Comment:
```suggestion
	/// Builds an [`UnsignedInvoiceRequest`] after checking for valid semantics.
``` 
And maybe link to the `sign` method